Einsum decomposition

sorenlassen opened the pull request:
This PR implements ONNXEinsumOp verify(), inferShapes(), and decomposes it into a set of supported ops (Constant, Squeeze, Unsqueeze, Mul, ReduceSum, Where, Transpose).

Closes #1367 

It is based on a preexisting python implementation here:
https://github.com/sorenlassen/onnx-fun/blob/main/python/src/einsummer.py

## Testing:

Added the onnx einsum tests to `inference_backend.py` and ran them with the command:
```TEST_VERBOSE=true TEST_CASE_BY_USER="test_einsum_batch_diagonal_cpu test_einsum_batch_matmul_cpu test_einsum_inner_prod_cpu test_einsum_transpose_cpu test_einsum_sum_cpu" cmake --build . --target check-onnx-backend```

Added shape inference tests in `onnx_shape_inference.mlir`

Added a unit test which can be run with:
```ctest -R TestONNXEinsumOp --verbose```

tungld commented:
@jenkins-droid test this please

tungld commented:
@sorenlassen thank you for the patch!

Could you please sign DCO by following its instruction (click Details by DCO)? To avoid this, we often sign our commits by `git -s`.
For the clang-format issue, you can call `git clang-format` locally to format the code. Thanks!

sorenlassen commented:
oops, I DCO signed the last commits now

I also ran clang-format just now

tungld commented:
> oops, I DCO signed the last commits now

I think we need to sign all the previous commits.

sorenlassen commented:
> > oops, I DCO signed the last commits now
> 
> I think we need to sign all the previous commits.

I had signed the initial batch of commits yesterday but hadn't signed the follow-on commits today

when I ran the DCO suggested command `git rebase HEAD~22 --signoff` is also signed the MHLO commit:
```https://github.com/sorenlassen/onnx-mlir-einsum/commit/d2a0cf37140ec032c6c741b923dd2b1c4afbad9f```
which looked wrong, so I manually scrubbed my signoff on that

shall I sign it anyway to make DCO happy?

sorenlassen commented:
> shall I signoff on the MHLO commit to make the DCO check happy?

I rebased and went ahead and signed off on  other people's MHLO and SeqType/LoopOp commits, hopefully this is all right and will pass the DCO check

sorenlassen commented:
> > shall I signoff on the MHLO commit to make the DCO check happy?
> 
> I rebased and went ahead and signed off on other people's MHLO and SeqType/LoopOp commits, hopefully this is all right and will pass the DCO check

should be fixed properly now: I dropped the MHLO and SeqType/LoopOp commits and then rebased cleanly against upstream/main

tungld commented:
@jenkins-droid test this please

tungld commented on src/Transform/ONNX/DecomposeEinsum.cpp:
```diff
@@ -0,0 +1,591 @@
+    output.value = builder
+                       .create<ONNXTransposeOp>(loc, output.type(elementType),
+                           output.value, builder.getI64ArrayAttr(perm))
+                       .getResult();
```
We are recently moving to use DialectBuilder instead of building an op directly, which is easier to maintain our code, e.g.:
```
MultiDialectBuilder<OnnxBuilder> create(buider, loc);
create.onnx.transpose(...)
```
The DialectBuilder for ONNX is here: https://github.com/onnx/onnx-mlir/blob/main/src/Dialect/ONNX/DialectBuilder.hpp. Could you please use DialectBuilder instead? and feel free to add new operations if they do not exist.

sorenlassen replied:
> Could you please use DialectBuilder instead? and feel free to add new operations if they do not exist.

OK. So far I made use of DialectBuilder for the ops that it defines. I'll try to add ops I need that don't exist, e.g. Squeeze, ReduceSum.

I ran into a problem with matmul() which returns toMemRef() which breaks things for me. Should I add a variant of matmul() that omits toMemRef()?

tungld replied:
>  Should I add a variant of matmul() that omits toMemRef()?

Yes, please. I think it is a worth variant. Thanks!

sorenlassen replied:
Since the call to toMemref() seemed non-standard, I renamed the old matmul() with toMemref() and changed the existing callers, and then I made plain matmul() not call toMemref() and put that to use in DecomposeEinsum:
https://github.com/onnx/onnx-mlir/pull/1537/commits/9fd808927d0f4ed818ca9408b146f67507a96b76

I'm happy to change it back of course if you wish, just let me know.

tungld replied:
How about explicitly calling `toMemref` when needed, and removing `matmulToMemref` e.g.
```
Value XtWT =
      create.onnx.toMemref(create.onnx.matmul(matrixAllGatesType, Xt, weightPack.WT));
```

sorenlassen replied:
> How about explicitly calling toMemref when needed, and removing matmulToMemref

good idea, done

tungld replied:
Thank you so much!

tungld commented on src/Transform/ONNX/DecomposeEinsum.cpp:
```diff
@@ -0,0 +1,591 @@
+      return;
+    }
+    assert(ShapedType::getNumElements(output.shape) ==
+           ShapedType::getNumElements(reShape));
```
Please add an assertion message.

sorenlassen replied:
done

tungld commented on src/Transform/ONNX/DecomposeEinsum.cpp:
```diff
@@ -0,0 +1,591 @@
+           ShapedType::getNumElements(reShape));
+    // no zero dims (dispatched at the beginning in decompose()) so
+    // we can ignore ONNXReshapeOp allowzero
+    assert(ShapedType::getNumElements(reShape) > 0);
```
Please add an assertion message.

sorenlassen replied:
done

tungld commented on test/backend/inference_backend.py:
```diff
@@ -283,6 +283,11 @@ def get_test_models():
         # ==OP== DynamicQuantizeLinear
 
         # ==OP== EinSum
+        "test_einsum_batch_diagonal_cpu": {STATIC_SHAPE:{}, DYNAMIC_SHAPE:{-1:{-1}}, CONSTANT_INPUT:{-1}},
```
Could you please add a magic line (starting with `# ==LIM==`) about type limitation at L286, e.g.
```
# ==LIM== Limited to the types supported by ReduceSum and MatMul (which we decompose to in most cases) which exclude integers with width < 32
```
This magic line will be used to generate this document: https://github.com/onnx/onnx-mlir/blob/main/docs/SupportedONNXOps-cpu.md. You will find a python command to generate the document at the beginning of `SupportedONNXOps-cpu.md`. Please run the python command to re-generate the document.

tungld reviewed (commented):
This is outstanding. Thank you so much for your hard work!

sorenlassen commented on src/Dialect/ONNX/ONNXEinsumOpHelper.cpp:
```diff
@@ -0,0 +1,319 @@
+FailureOr<Signature> inferSignature(
+    ONNXEinsumOpAdaptor operandAdaptor, ErrorFn emitErrorFn) {
+  auto equation = operandAdaptor.equation();
+  auto inputs = operandAdaptor.Inputs();
```
done

I only left auto when it helps avoid writing out long-winded iterator and ArrayRef types

sorenlassen commented on src/Dialect/ONNX/ONNXEinsumOpHelper.cpp:
```diff
@@ -0,0 +1,319 @@
+  }
+  SmallVector<StringRef> equationInputs;
+  commaSeparatedInputs.split(equationInputs, ',');
+  assert(equationInputs.size() == inputs.size()); // precondition
```
done

sorenlassen commented on src/Dialect/ONNX/ONNXEinsumOpHelper.cpp:
```diff
@@ -0,0 +1,319 @@
+  assert(succeeded(verifyEquation(equation, inputs.size(),
+      emitErrorFn))); // precondition, TODO: remove this excessive check
+  assert(llvm::all_of(inputs, [](Value i) {
+    return i.getType().cast<ShapedType>().hasRank();
```
I used isRankedShapedType from TypeUtilities, thanks for the suggestion

FWIW, I gather it isn't really needed to check if types are ShapedType here because the EInsum spec says that all types are tensors and that's checked in ONNXEinsumOp::verifyInvariants()

tungld commented:
@jenkins-droid test this please

tungld commented:
@jenkins-droid test this please

tungld commented:
@jenkins-droid test this please

tungld commented:
@jenkins-droid test this please

tungld commented:
@sorenlassen could you please check why the lit test in  `onnx_shape_inference_error_einsum.mlir` failed? Thanks!

sorenlassen commented:
> @sorenlassen could you please check why the lit test in `onnx_shape_inference_error_einsum.mlir` failed? Thanks!

it succeeds for me locally, as far as I can tell

is the CI failing? how can I see the failure?

tungld commented:
> is the CI failing? how can I see the failure?

Yes, can you view the CI logs when clicking `Details` button by each Jenkins CI?

Here is the log:
```
[100%] Built target onnx-mlir
Scanning dependencies of target check-onnx-lit
[100%] Running the ONNX-MLIR regression tests
-- Testing: 152 tests, 32 workers --
Testing:  0.. 10.. 20.. 30.. 40.. 50.. 60.. 70.. 80.. 90.
FAIL: Open Neural Network Frontend :: onnx/onnx_shape_inference_error_einsum.mlir (149 of 152)
******************** TEST 'Open Neural Network Frontend :: onnx/onnx_shape_inference_error_einsum.mlir' FAILED ********************
Script:
--
: 'RUN: at line 1';   /workdir/onnx-mlir/build/Debug/bin/onnx-mlir-opt --shape-inference /workdir/onnx-mlir/test/mlir/onnx/onnx_shape_inference_error_einsum.mlir -split-input-file -verify-diagnostics
--
Exit Code: 139

Command Output (stderr):
--
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.	Program arguments: /workdir/onnx-mlir/build/Debug/bin/onnx-mlir-opt --shape-inference /workdir/onnx-mlir/test/mlir/onnx/onnx_shape_inference_error_einsum.mlir -split-input-file -verify-diagnostics
 #0 0x00005565ab5350c4 PrintStackTraceSignalHandler(void*) Signals.cpp:0:0
 #1 0x00005565ab532864 SignalHandler(int) Signals.cpp:0:0
 #2 0x00007f31720e3420 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x14420)
 #3 0x00005565ab4010e5 mlir::Operation::emitError(llvm::Twine const&) (/workdir/onnx-mlir/build/Debug/bin/onnx-mlir-opt+0x1d1e0e5)
 #4 0x00005565ab401c50 mlir::Operation::emitOpError(llvm::Twine const&) (/workdir/onnx-mlir/build/Debug/bin/onnx-mlir-opt+0x1d1ec50)
 #5 0x00005565ab402945 mlir::OpState::emitOpError(llvm::Twine const&) (/workdir/onnx-mlir/build/Debug/bin/onnx-mlir-opt+0x1d1f945)
 #6 0x00005565a98fc5d0 mlir::ONNXEinsumOp::verify()::'lambda'()::operator()() const /workdir/onnx-mlir/src/Dialect/ONNX/ONNXOps.cpp:4021:35
 #7 0x00005565a9ab4c45 mlir::InFlightDiagnostic llvm::function_ref<mlir::InFlightDiagnostic ()>::callback_fn<mlir::ONNXEinsumOp::verify()::'lambda'()>(long) /workdir/llvm-project/llvm/include/llvm/ADT/STLFunctionalExtras.h:47:3
 #8 0x00005565a9ca8378 llvm::function_ref<mlir::InFlightDiagnostic ()>::operator()() const /workdir/llvm-project/llvm/include/llvm/ADT/STLFunctionalExtras.h:69:3
 #9 0x00005565a9ca48f9 onnx_mlir::einsum::verifyEquation(llvm::StringRef, unsigned long, llvm::function_ref<mlir::InFlightDiagnostic ()>) /workdir/onnx-mlir/src/Dialect/ONNX/ONNXEinsumOpHelper.cpp:55:29
#10 0x00005565a98fc768 mlir::ONNXEinsumOp::verify() /workdir/onnx-mlir/src/Dialect/ONNX/ONNXOps.cpp:4027:13
#11 0x00005565a9b89d60 mlir::Op<mlir::ONNXEinsumOp, mlir::OpTrait::ZeroRegion, mlir::OpTrait::OneResult, mlir::OpTrait::OneTypedResult<mlir::Type>::Impl, mlir::OpTrait::ZeroSuccessor, mlir::OpTrait::VariadicOperands, mlir::OpTrait::OpInvariants, mlir::MemoryEffectOpInterface::Trait, mlir::ShapeInference::Trait>::verifyInvariants(mlir::Operation*) /workdir/llvm-project/mlir/include/mlir/IR/OpDefinition.h:1851:15
#12 0x00005565a98bb5d4 mlir::LogicalResult llvm::detail::UniqueFunctionBase<mlir::LogicalResult, mlir::Operation*>::CallImpl<mlir::LogicalResult (* const)(mlir::Operation*)>(void*, mlir::Operation*) /workdir/llvm-project/llvm/include/llvm/ADT/FunctionExtras.h:223:3
#13 0x00005565ab42d6c9 (anonymous namespace)::OperationVerifier::verifyOperation(mlir::Operation&) Verifier.cpp:0:0
#14 0x00005565ab42dd3c (anonymous namespace)::OperationVerifier::verifyOperation(mlir::Operation&) Verifier.cpp:0:0
#15 0x00005565ab42e8b5 (anonymous namespace)::OperationVerifier::verifyOperation(mlir::Operation&) Verifier.cpp:0:0
#16 0x00005565ab42fbea mlir::verify(mlir::Operation*, bool) (/workdir/onnx-mlir/build/Debug/bin/onnx-mlir-opt+0x1d4cbea)
#17 0x00005565ab314b6c mlir::parseSourceFile(llvm::SourceMgr const&, mlir::Block*, mlir::MLIRContext*, mlir::LocationAttr*, mlir::AsmParserState*) (/workdir/onnx-mlir/build/Debug/bin/onnx-mlir-opt+0x1c31b6c)
#18 0x00005565aa1698bd performActions(llvm::raw_ostream&, bool, bool, llvm::SourceMgr&, mlir::MLIRContext*, llvm::function_ref<mlir::LogicalResult (mlir::PassManager&)>) (.constprop.0) MlirOptMain.cpp:0:0
#19 0x00005565aa16a3fb processBuffer(llvm::raw_ostream&, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer> >, bool, bool, bool, bool, llvm::function_ref<mlir::LogicalResult (mlir::PassManager&)>, mlir::DialectRegistry&, llvm::ThreadPool*) MlirOptMain.cpp:0:0
#20 0x00005565aa16a95c mlir::LogicalResult llvm::function_ref<mlir::LogicalResult (std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer> >, llvm::raw_ostream&)>::callback_fn<mlir::MlirOptMain(llvm::raw_ostream&, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer> >, llvm::function_ref<mlir::LogicalResult (mlir::PassManager&)>, mlir::DialectRegistry&, bool, bool, bool, bool, bool)::'lambda'(std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer> >, llvm::raw_ostream&)>(long, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer> >, llvm::raw_ostream&) MlirOptMain.cpp:0:0
#21 0x00005565ab445fad mlir::splitAndProcessBuffer(std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer> >, llvm::function_ref<mlir::LogicalResult (std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer> >, llvm::raw_ostream&)>, llvm::raw_ostream&) (/workdir/onnx-mlir/build/Debug/bin/onnx-mlir-opt+0x1d62fad)
#22 0x00005565aa16aaf9 mlir::MlirOptMain(llvm::raw_ostream&, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer> >, mlir::PassPipelineCLParser const&, mlir::DialectRegistry&, bool, bool, bool, bool, bool) (/workdir/onnx-mlir/build/Debug/bin/onnx-mlir-opt+0xa87af9)
#23 0x00005565a97f3967 main /workdir/onnx-mlir/src/Tools/onnx-mlir-opt/onnx-mlir-opt.cpp:168:16
#24 0x00007f3171bb5083 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24083)
#25 0x00005565a97f21be _start (/workdir/onnx-mlir/build/Debug/bin/onnx-mlir-opt+0x10f1be)
/workdir/onnx-mlir/build/test/mlir/onnx/Output/onnx_shape_inference_error_einsum.mlir.script: line 1:  6707 Segmentation fault      (core dumped) /workdir/onnx-mlir/build/Debug/bin/onnx-mlir-opt --shape-inference /workdir/onnx-mlir/test/mlir/onnx/onnx_shape_inference_error_einsum.mlir -split-input-file -verify-diagnostics

--
```

sorenlassen commented:
@tungld thanks for the info

It passed for me locally and I can't tell why it fails in the CO. It wasn't important anyway, so I removed it now.

tungld commented:
@jenkins-droid test this please

tungld reviewed (approved):
LGTM. Thank you very much!

sorenlassen commented:
> LGTM. Thank you very much!

thanks for the review!

could you merge it for me?

tungld commented:
> could you merge it for me?

Yes, sure. I am waiting for comments from other reviewers if any. Otherwise, I will merge it tomorrow.
